### PR TITLE
Adamax/Nadam/AMSGrad optimizers

### PR DESCRIPTION
**Pull request by yura-ata**
Couple of open points:
* AMSGrad needs 3 meta args - which looks awkwaed using zip/unzip
* Note sure about extracting `unbias` into MathOp
* Missing scaladocs

### CODE REVIEW

File: src/main/scala/org/scanet/math/MathBaseOp.scala [hunk 1]
```diff
@@ -131,23 +131,127 @@ import scala.Ordering.Implicits._
...
+   * val iter = 2.const
+   * x.unbiased(bias, iter).eval should be(Tensor.scalar(2))
+   * }}}
+   *
+   * @param out tensor
+   * @param bias biased component
+   * @param iter computation iteration
+   * @return unbiased value
+   */
+  def unbiased[A: TensorType: Numeric](out: F[A], bias: F[A], iter: F[Int]): F[A]
```
Comment (pashashiz) [inline src/main/scala/org/scanet/math/MathBaseOp.scala]: That expression is really fixing the fact that the algorithm is `0` biased at first steps and slowly accelerates. But the function itself has nothing to do with `unbiasing`. I would suggest renaming that to something like `boost` which increases the value when `iter` is small and when iter is reaching infinity that function works like identity.

File: src/main/scala/org/scanet/optimizers/AMSGrad.scala [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    val correctedVelocity = Tensor.zeros[Float](shape).const
+    val suffix = Tensor.zeros[Float](shape).const
+    ((momentum zip velocity) zip (correctedVelocity zip suffix)).eval
+  }
+
+  override def delta(grad: Output[Float], meta: Output[Float], iter: Output[Int]): Delta = {
+    val (args1, args2) = meta.unzip
+    val (prevMomentum, prevVelocity) = args1.unzip
+    val (prevCorrectedVelocity, suffix) = args2.unzip
+    val momentum = prevMomentum.decayingAvg(grad, beta1)
```
Comment (pashashiz) [inline src/main/scala/org/scanet/optimizers/AMSGrad.scala]: Could you rename it to`m` which stands for `mean`?

File: src/main/scala/org/scanet/optimizers/AMSGrad.scala [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    val velocity = prevVelocity.decayingAvg(grad.sqr, beta2)
```
Comment (pashashiz) [inline src/main/scala/org/scanet/optimizers/AMSGrad.scala]: Could you rename it to`v` which stands for `variance`? (for all such algs)

File: src/main/scala/org/scanet/optimizers/AMSGrad.scala [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+import org.scanet.syntax._
+
+case class AMSGrad(rate: Output[Float], beta1: Output[Float], beta2: Output[Float], epsilon: Output[Float]) extends Algorithm {
+
+  override def initMeta(shape: Shape): Tensor[Float] = {
+    val momentum = Tensor.zeros[Float](shape).const
+    val velocity = Tensor.zeros[Float](shape).const
+    val correctedVelocity = Tensor.zeros[Float](shape).const
+    val suffix = Tensor.zeros[Float](shape).const
+    ((momentum zip velocity) zip (correctedVelocity zip suffix)).eval
```
Comment (pashashiz) [inline src/main/scala/org/scanet/optimizers/AMSGrad.scala]: Could you add a new `zip` method so we would get:
```
zip(momentum, velocity, correctedVelocity)
```

File: src/main/scala/org/scanet/optimizers/AMSGrad.scala [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+  }
+
+  override def delta(grad: Output[Float], meta: Output[Float], iter: Output[Int]): Delta = {
+    val (args1, args2) = meta.unzip
+    val (prevMomentum, prevVelocity) = args1.unzip
```
Comment (pashashiz) [inline src/main/scala/org/scanet/optimizers/AMSGrad.scala]: Maybe `unzip3`...

File: src/main/scala/org/scanet/optimizers/Adamax.scala [hunk 6]
```diff
@@ -0,0 +1,26 @@
+package org.scanet.optimizers
+
+import org.scanet.core._
+import org.scanet.syntax._
+
+case class Adamax(rate: Output[Float], beta1: Output[Float], beta2: Output[Float], epsilon: Output[Float]) extends Algorithm {
```
Comment (pashashiz) [inline src/main/scala/org/scanet/optimizers/Adamax.scala:19]: Could you add a doc from https://keras.io/api/optimizers/adamax ?

File: src/main/scala/org/scanet/optimizers/AMSGrad.scala [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    val (prevCorrectedVelocity, suffix) = args2.unzip
```
Comment (pashashiz) [inline src/main/scala/org/scanet/optimizers/AMSGrad.scala]: Maybe we can name `variance` variable as:
- `vPrev` for previous `variance`
- `vCurrent` for current `variance` 
- `v` result `variance` which is output of `max` functinon

File: src/main/scala/org/scanet/optimizers/Nadam.scala [hunk 8]
```diff
@@ -0,0 +1,27 @@
+package org.scanet.optimizers
+
+import org.scanet.core._
+import org.scanet.syntax._
+
+case class Nadam(rate: Output[Float], beta1: Output[Float], beta2: Output[Float], epsilon: Output[Float]) extends Algorithm {
```
Comment (pashashiz) [inline src/main/scala/org/scanet/optimizers/Nadam.scala]: Could you also rename:
- `board/AdagGrad` to `board/AdaGrad`
- `board/AdagDelta` to `board/AdaDelta`
I made a typo...

**Comment by yura-ata** [comment]
* added `zip`(3) and `unzip3`
* added missing scaladocs
* renamed `unbias` to `boost`
* used more conventional names for some variables

**Review by pashashiz — APPROVED** [review]
Awesome!

**Pull request merged**